### PR TITLE
Bottom page

### PR DESCRIPTION
**Pull request by Scooletz**
This PR introduces a `BottomPage` that allows accumulating a lot of data changes before flushing them out as a `DataPage`. To make it work a `BottomPage` provides two pointers `Left` and `Right` that allow to divide keys to two buckets only. Then, on the next level, grandchildren are created if needed. This allows right now to create 7 bottom pages (1 on the 0th level, 2 on the 1st level, 4 on the 3rd level) which greatly reduces greedy fan outs. This could be pushed further with the next level and some more overflow logic if needed.

The structure can be optimized to be used to roll up `DataPage` on delete. At the same time such big deletes seem to be unlikely. 

Similar to #387

### Benchmarks

Database size reduced to `269GB`, comparing to `RocksDb` of Nethermind with `158GB`. Access times seem to be not impacted at all. Usually it can be up to 1 more level to query.

image

image

### CODE REVIEW

File: src/Paprika/Store/DataPage.cs [hunk 1]
```diff
@@ -129,187 +90,86 @@ private static void Set(DbAddress at, in NibblePath key, in ReadOnlySpan<byte> d
...
 
-                // Try set again
-                if (map.TrySet(k, data))
-                {
-                    // Update happened, return
-                    break;
-                }
+            // Create a child
+            var lvl = (byte)(page.Header.Level + ConsumedNibbles);
+            batch.GetNewPage<BottomPage>(out childAddr, lvl);
```
Comment (dipkakwani) [inline src/Paprika/Store/DataPage.cs:160]: I might be completely wrong here (or it might be already existing and I don't see it :P) : While allocating a new BottomPage, would it be useful to do some form of bulk page allocation? Most BottomPages would eventually need to allocate its children and having the children pages contiguous in the memory would make most operations faster?

**Review by dipkakwani — APPROVED** [review]

File: src/Paprika.Tests/Store/DataPageTests.cs [hunk 2]
```diff
@@ -20,9 +20,7 @@ public class DataPageTests : BasePageTests
     public void Spinning_through_same_keys_should_use_limited_number_of_pages()
     {
         var batch = NewBatch(BatchId);
-        var page = batch.GetNewPage(out _, true);
-
-        var data = new DataPage(page);
+        var data = ((IBatchContext)batch).GetNewPage<DataPage>(out _);
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Paprika.Tests/Store/DataPageTests.cs:23]: Oh. The reason was a major rewrite that happend in #392 and a few others. Let me delete them in a separate PR and we can start adding them as needed.

File: src/Paprika/Store/DataPage.cs (hunk 1 above)
Comment (Scooletz) [reply src/Paprika/Store/DataPage.cs:160]: > Most BottomPages would eventually need to allocate its children.

They might have some children depending on the data but don't have to. We could allocate more of them but this would allocate empty space and we don't want to over allocate. With the current approach, we allocate only when there's something to flush down to a given page. This ensures that we save up a lot.

> having children pages contiguous in the memory would make most operations faster

Paprika uses `COW`. What not is contiguous, probably will not be after a few COWs. Imagine the root being COWed in one batch, then the child in next. A probability that they will be _contiguous_ is quite unlikely I believe. One thing to mention is that we do sort abandoned pages and serve them that way. We also allocate from the end when no abandoned can be reused. This should increase the probability of pages allocated one after another and as we COW as we walk through the pages... maybe we get a higher chance of continuity for free?

File: src/Paprika.Runner/StatisticsForPagedDb.cs [hunk 3]
```diff
@@ -77,12 +77,11 @@ private static Layout BuildReport(StatisticsVisitor.Stats stats, string name, in
...
-        t.AddColumn(new TableColumn("Page count"));
-        t.AddColumn(new TableColumn("Leaf page count"));
-        t.AddColumn(new TableColumn("Overflow page count"));
-        t.AddColumn(new TableColumn($"{nameof(DataPage)}-inner (P50) % usage"));
-        t.AddColumn(new TableColumn($"{nameof(DataPage)}-leaf (P50) % usage"));
-        t.AddColumn(new TableColumn($"{nameof(LeafOverflowPage)} (P50) % usage"));
+        t.AddColumn(new TableColumn("Total page count"));
+        t.AddColumn(new TableColumn("Data page count"));
+        t.AddColumn(new TableColumn("Bottom page count"));
+        t.AddColumn(new TableColumn($"{nameof(DataPage)} P50 % usage"));
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Paprika.Runner/StatisticsForPagedDb.cs:83]: I'm not sure I'm following. This is a static view that you can run with `Cli stats` (added it to the description as additional picture). If we're talking about metrics we do aggregate some of them with `System.Diagnostics.Metrics`. We don't gather percentage used etc as it would be hard (update every time?). I think I'm not following the comment fully. Could you rephrase?

File: src/Paprika/Store/DbAddressList.cs [hunk 4]
```diff
@@ -126,6 +126,7 @@ private static DbAddress[] ToArrayImpl<TList>(in TList list)
         return array;
     }
 
+
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Paprika/Store/DbAddressList.cs:129]: Maybe in a separate PR with the tests.

File: src/Paprika.Runner/StatisticsForPagedDb.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dipkakwani) [reply src/Paprika.Runner/StatisticsForPagedDb.cs:83]: Hmm, got it - this is just the static view once the node is fully synced. What I meant was that could this static view be captured across time intervals (say every few mins?). I agree that it would be compute intensive, but would that data be useful to understand the trend/correlation at different times?
